### PR TITLE
Fix bad statuses reported in ssbc

### PR DESCRIPTION
**Pull request by eseliger**
This fixes two issues:
- Cached workspaces incorrectly appeared when filtered by "PENDING"
- Cached workspaces incorrectly were ignored in the stats bar, they are complete now.

Closes https://github.com/sourcegraph/sourcegraph/issues/35058
Closes https://github.com/sourcegraph/sourcegraph/issues/35061

## Test plan

Adjusted test suites and manually verified the issues from the tickets are now gone.

### CODE REVIEW

File: enterprise/internal/batches/store/batch_specs_test.go [hunk 1]
```diff
@@ -656,6 +666,7 @@ func TestStoreGetBatchSpecStats(t *testing.T) {
 			Canceling:         1,
 			Canceled:          1,
 			Failed:            1,
+			CachedWorkspaces:  1,
```
[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/internal/batches/store/batch_specs_test.go:669]: This internal struct doesn't work like the one we expose in the resolver layer, it separately counts finished+cached and executed+successful. We need that internally to compute the batch spec state. That's why we are adding the two up above in the resolver layer code.

[the remark this answers is not in the document]
Comment (courier-new) [reply enterprise/internal/batches/store/batch_specs_test.go:669]: Oh thank you, I was definitely conflating the two. 👍

**Review by courier-new — APPROVED** [review]

**Comment by adeola-ak** [comment]
how does a workspace become cached?

**Comment by eseliger** [comment]
When there was an execution with the exact same parameters, steps, and such before and we still hold on to the result of that, we don't need to re-execute, because we can recycle the execution result. The cache is read in the workspace resolution worker.

**Comment by adeola-ak** [comment]
ok got it, thank you!

**Pull request merged**